iree/compiler/Bindings/TFLite/Transforms/MaterializeShapeSupport.cpp

```diff
@@ -0,0 +1,462 @@
+    // needed. This is only required if there are dynamic shapes.
+    auto dirtyVarOp = moduleBuilder.create<IREE::Flow::VariableOp>(
+        loc, funcOp.getName().str() + "_shapes_dirty",
+        /*isMutable=*/true, moduleBuilder.getI1Type(),
+        moduleBuilder.getIntegerAttr(moduleBuilder.getI1Type(), 1));
+    dirtyVarOp.setPrivate();
+    auto calculateShapeFuncOp =
+        createShapeCalculationFunc(loc, namePrefix, funcOp, inputVarOps,
+                                   outputVarOps, dirtyVarOp, moduleBuilder);
+
+    // Create input query function (just reads variables).
+    createQueryInputShapeFunc(loc, namePrefix, inputVarOps, moduleBuilder);
+
+    // Create input resize function (updates variables, set dirty flag).
+    createResizeInputShapeFunc(loc, namePrefix, inputVarOps, dirtyVarOp,
+                               moduleBuilder);
+
+    // Create output query function (if dirty recalculates shapes).
+    createQueryOutputShapeFunc(loc, namePrefix, outputVarOps,
+                               calculateShapeFuncOp, moduleBuilder);
+
+    return success();
+  }
+
+  // Creates and initializes to default values one flow.variable for each I/O
+  // shape of the given |funcOp|. |inputVarOps| and |outputVarOps| will be
+  // populated with the created variables.
+  void createShapeVariables(
+      Location loc, StringRef namePrefix, FuncOp funcOp,
+      SmallVectorImpl<IREE::Flow::VariableOp> &inputVarOps,
+      SmallVectorImpl<IREE::Flow::VariableOp> &outputVarOps,
+      OpBuilder &moduleBuilder) {
+    auto funcType = funcOp.getType();
+
+    // TFLite requires the tensor names at runtime. If they've previously been
+    // extracted into iree.identifiers we use those and otherwise fallback to
+    // a generic naming scheme that matches the IR (somewhat).
+    SmallVector<std::string, 4> inputNames;
+    SmallVector<std::string, 4> outputNames;
+    for (unsigned i = 0; i < funcType.getNumInputs(); ++i) {
+      auto identifier =
+          funcOp.getArgAttrOfType<StringAttr>(i, "iree.identifier");
+      if (identifier) {
+        inputNames.push_back(identifier.getValue().str());
+      } else {
+        inputNames.push_back(std::string("arg") + std::to_string(i));
+      }
+    }
+    for (unsigned i = 0; i < funcType.getNumResults(); ++i) {
+      auto identifier =
+          funcOp.getResultAttrOfType<StringAttr>(i, "iree.identifier");
+      if (identifier) {
+        outputNames.push_back(identifier.getValue().str());
+      } else {
+        outputNames.push_back(std::string("ret") + std::to_string(i));
+      }
+    }
+
+    for (auto input : llvm::zip(inputNames, funcType.getInputs())) {
+      auto name = funcOp.getName().str() + "_" + std::get<0>(input) + "_shape";
+      auto type = std::get<1>(input);
+      auto tensorType = type.dyn_cast<TensorType>();
+      assert(tensorType && "expecting only tensors in tflite function I/O");
```

Comment:
Maybe later on we could add a validation pass at the front of the import that checks for conditions like this, to give slightly more organized error messages?

user: "Convert my model for running via the TF Lite bindings"
tool: "Your model uses features [x, y, z] that the bindings requested do not support. Consider using the foo bindings or C API, or changing your model."

Reply:
Good idea - and it can also check for features that are missing (like variables/etc) - though some of them (like data dependent dynamic shapes) can't easily be checked for without more lowering. Filed #4710 to track.